### PR TITLE
Add Bookkeeping

### LINKED ISSUE #34: Scytale should log a Bookkeeping Entry like Tr1d1um
Scytale should log a Bookeeping Entry like Tr1d1um.

https://github.com/Comcast/tr1d1um/blob/46607c2663aae15358ed34d58e5dcdeecf891466/src/tr1d1um/common/utils.go#L35

Data to log:
```
Wrp.msg_type
Wrp.dest
Wrp.source
Wrp.transaction_uuid
Wrp.status
Wrp.partner_ids
satClientID
```

### PR DESCRIPTION
**Pull request by kcajmagic**
- Added Bookkeeper to handle logging for different routes. 

Closes #34

### CODE REVIEW

File: src/scytale/primaryHandler.go [hunk 1]
```diff
@@ -132,21 +145,6 @@ func NewPrimaryHandler(logger log.Logger, v *viper.Viper, registry xmetrics.Regi
 				logginghttp.SetLogger(
 					logger,
 					logginghttp.RequestInfo,
-
-					// custom logger func that extracts the intended destination of requests
-					func(kv []interface{}, request *http.Request) []interface{} {
```
Comment (njharter) [inline src/scytale/primaryHandler.go]: why are you removing this block of code?

**Comment by schmidtw** [comment]
@njharter can this be merged?

**Comment by johnabass** [comment]
We don't want to be doing bookkeeping this way, in any of our servers:

- Bookkeeping should not be present in application layer code.  This makes it hard to refactor later.
- Bookkeeping is a cross-cutting concern, and should be done with an Alice-style decorator.  There is already code to facilitate this in the `logging` and `xcontext` packages.
- Logging decorations should transfer to every log message in an HTTP code path, not just the bookkeeping message.

**Comment by johnabass** [comment]
There are (3) basic steps to a better solution.  The first (2) need only be done once, for all servers.

(1) Use `xcontext.Populate` to create an Alice-style decorator:
```
c := xcontext.Populate(0, logginghttp.SetLogger(
        logginghttp.Header("X-Webpa-Device-Name", "deviceName"),
        // any other go-kit request functions that you want go here, e.g. to pull data from a SAT
    ),
)
```

(2) Create a bookkeeping decorator, preferably in one place such as the `logginghttp` package:
```
b := func(next http.Handler) http.Handler {
    return http.HandlerFunc(func(response http.ResponseWriter, request *http.Request) {
        logging.GetLogger(request.Context()).Log(level.Key(), level.InfoValue(), logging.MessageKey(), "Bookkeeping")
    })
}
```

(3) Use both of the above decorators for any handlers in application code:
```
r := mux.NewRouter()

// decorate one handler
r.Handle("/foobar", alice.New(c, b).Then(myHandler))

// decorate the whole thing
alice.New(c, b).Then(r)
```

File: src/scytale/primaryHandler.go [hunk 2]
```diff
@@ -310,3 +274,48 @@ func NewPrimaryHandler(logger log.Logger, v *viper.Viper, registry xmetrics.Regi
 
 	return router, nil
 }
+
+func doFanout(ctx context.Context, fanout *http.Request, message *wrp.Message) (context.Context, error) {
+	populateMessage(ctx, message)
+
+	bookkeepingLog(ctx, message)
```
Comment (johnabass) [inline src/scytale/primaryHandler.go]: We don't want this here.  This is in application-layer code.  We want the bookkeeping done as part of an Alice-style decorator.  This decorator can go into the `logginghttp` package as a standard thing applications can add to their handler mappings.

File: src/scytale/primaryHandler.go [hunk 3]
```diff
@@ -310,3 +274,48 @@ func NewPrimaryHandler(logger log.Logger, v *viper.Viper, registry xmetrics.Regi
...
+
+	fanoutBody := buffer.Bytes()
+	fanout.Body, fanout.GetBody = xhttp.NewRewindBytes(fanoutBody)
+	fanout.ContentLength = int64(len(fanoutBody))
+	fanout.Header.Set("Content-Type", wrp.Msgpack.ContentType())
+	fanout.Header.Set("X-Webpa-Device-Name", message.Destination)
+	return ctx, nil
+}
+
+func bookkeepingLog(ctx context.Context, message *wrp.Message) error {
```
Comment (johnabass) [inline src/scytale/primaryHandler.go]: This function should not exist in application code.  What you want is decoration:

```
package logginghttp

import (
    "github.com/Comcast/webpa-common/logging"
    "net/http"
)

func Bookkeeper(next http.Handler) http.Handler {
    return http.HandlerFunc(func(response http.ResponseWriter, request *http.Request) {
        // rely on another decorator, e.g. xcontext.Populate, to supply the fields
        logging.GetLogger(request.Context()).Log(logging.MessageKey(), "Bookkeeping")
        next.ServeHTTP(response, request)
    })
}
```

Comment (johnabass) [reply src/scytale/primaryHandler.go]: Client code, such as in `scytale`, will then supply a contextual logger, again via decoration:

```
// imports omitted
c := xcontext.Populate(0,
    logginghttp.SetLogger(loggerFromViper, logginghttp.Header("X-Example", "example")),
)

// use 'c' to decorate handlers:
alice.New(c, /* other constructors).Then(myHandler)
```

**Pull request closed without merging**